### PR TITLE
fix rasterio chunking with s3 datasets

### LINKED ISSUE #1816: rasterio chunks argument causes loading from s3 to fail
#### Code Sample, a copy-pastable example if possible

```python
# This works
url = 's3://landsat-pds/L8/139/045/LC81390452014295LGN00/LC81390452014295LGN00_B1.TIF'
ds = xr.open_rasterio(url)
# this doesn't
ds = xr.open_rasterio(url, chunks=512)
```

The error is
```
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
<ipython-input-17-8b55d7e920b8> in <module>()
      6 # https://aws.amazon.com/public-datasets/landsat/
      7 # 512x512 chunking
----> 8 ds = xr.open_rasterio(url, chunks=512)
      9 ds

~/miniconda3/envs/geo_scipy/lib/python3.6/site-packages/xarray-0.10.0-py3.6.egg/xarray/backends/rasterio_.py in open_rasterio(filename, chunks, cache, lock)
    172         from dask.base import tokenize
    173         # augment the token with the file modification time
--> 174         mtime = os.path.getmtime(filename)
    175         token = tokenize(filename, mtime, chunks)
    176         name_prefix = 'open_rasterio-%s' % token

~/miniconda3/envs/geo_scipy/lib/python3.6/genericpath.py in getmtime(filename)
     53 def getmtime(filename):
     54     """Return the last modification time of a file, reported by os.stat()."""
---> 55     return os.stat(filename).st_mtime
     56 
     57 

FileNotFoundError: [Errno 2] No such file or directory: 's3://landsat-pds/L8/139/045/LC81390452014295LGN00/LC81390452014295LGN00_B1.TIF'
```

#### Problem description

It is pretty clear that the current xarray code expects to receive a filename. (The name of the argument is `filename`.) But rasterio's `open` function accepts a much wider range of [dataset identifiers](https://mapbox.github.io/rasterio/switch.html#dataset-identifiers). The tokenizing function should be updated to allow for this. Seems like it should be a pretty easy fix.

#### Output of ``xr.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.2.final.0
python-bits: 64
OS: Darwin
OS-release: 16.7.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

xarray: 0.10.0
pandas: 0.20.3
numpy: 1.13.1
scipy: 0.19.1
netCDF4: 1.3.1
h5netcdf: 0.4.1
Nio: None
bottleneck: 1.2.1
cyordereddict: None
dask: 0.16.0
matplotlib: 2.1.0
cartopy: 0.15.1
seaborn: 0.8.1
setuptools: 36.3.0
pip: 9.0.1
conda: None
pytest: 3.2.1
IPython: 6.1.0
sphinx: 1.6.5
</details>

### PR DESCRIPTION
**Pull request by rabernat**
- [x] Closes #1816 (remove if there is no corresponding issue, which should only be the case for minor changes)
 - [x] Tests added (for all bug fixes or enhancements)
 - [x] Tests passed (for all non-documentation changes)
 - [x] Passes ``git diff upstream/master **/*py | flake8 --diff`` (remove if you did not edit any Python files)
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API (remove if this change should not be visible to users, e.g., if it is an internal clean-up, or if this is part of a larger project that will be documented later)

This is a simple fix for token generation of non-filename targets for rasterio.

The problem is that I have no idea how to test it without actually hitting s3 (which requires boto and aws credentials).

### CODE REVIEW

File: xarray/backends/rasterio_.py [hunk 1]
```diff
@@ -217,7 +217,11 @@ def open_rasterio(filename, chunks=None, cache=None, lock=None):
     if chunks is not None:
         from dask.base import tokenize
         # augment the token with the file modification time
-        mtime = os.path.getmtime(filename)
+        try:
+            mtime = os.path.getmtime(filename)
+        except FileNotFoundError:
+            # the filename is probably and s3 bucket rather than a regular file
```
Comment (shoyer) [inline xarray/backends/rasterio_.py]: and -> an

File: xarray/backends/rasterio_.py [hunk 2]
```diff
@@ -217,7 +217,11 @@ def open_rasterio(filename, chunks=None, cache=None, lock=None):
     if chunks is not None:
         from dask.base import tokenize
         # augment the token with the file modification time
-        mtime = os.path.getmtime(filename)
+        try:
+            mtime = os.path.getmtime(filename)
+        except FileNotFoundError:
```
Comment (shoyer) [inline xarray/backends/rasterio_.py]: Python 2 only has the super-class `OSError`, so that should probably be used instead here.

**Comment by shoyer** [comment]
You can test this by mocking `os.path.getmtime`, e.g.,
```python
from unittest import mock  # but note that we use a special import for mock in xarray

with mock.patch('os.path.getmtime', side_effect=RuntimeError):
  os.path.getmtime('fooasdf.txt')
```

**Comment by rabernat** [comment]
I also refactored the rasterio test suite to eliminate a lot of boilerplate dataset creation and added tests for network urls. I'm not a geotiff expert, so please let me know if I have done anything silly.

File: xarray/tests/test_backends.py [hunk 3]
```diff
@@ -2469,6 +2371,33 @@ def test_ENVI_tags(self):
...
+                with xr.open_rasterio(tmp_file, chunks=(1, 2, 2)) as actual:
+                    import dask.array as da
+                    self.assertIsInstance(actual.data, da.Array)
+                    assert_allclose(actual, expected)
+
+    @network
+    def test_http_url(self):
+        # more examples urls here
+        # http://download.osgeo.org/geotiff/samples/
+        url = 'http://download.osgeo.org/geotiff/samples/made_up/ntf_nord.tif'
```
Comment (fmaussion) [inline xarray/tests/test_backends.py]: I forgot about the `@network` decorator, this can be quite useful for other tests as well...

**Comment by rabernat** [comment]
> this will mess with #1712 and #1740 quite a bit

I probably should have paid more attention to those PR's. My apologies.

Here is how I ended up refactoring the tests. I had to write a new test to catch the bug I discovered. I looked at the existing tests for guidance and saw lots of boilerplate. So rather than copying the boilerplate, I refactored it.

Let me know how I can help resolve this conflicts this may cause with the other open PRs.

**Comment by fmaussion** [comment]
https://github.com/pydata/xarray/pull/1740 is OK since the tests are still missing. My PR is a bit more problematic, but I can handle it. I would favor pushing yours forward now and I'll rebase my PR on top of it.

**Comment by shoyer** [comment]
@rabernat this now has some conflicts to resolve. Can you fix those and merge this? Thanks!

**Comment by rabernat** [comment]
I resolved conflicts. LGTM before it gets stale again!

Is it bad form to merge my own PR?

**Comment by fmaussion** [comment]
test_nodata and test_nodata_missing are tests which were added in https://github.com/pydata/xarray/pull/1740, they shouldn't be deleted by your PR. It's fine if you don't refactorize them (I'll need to have a look at all the tests anyway once all conflicting PRs are merged, see https://github.com/pydata/xarray/issues/1843)

**Comment by shoyer** [comment]
@rabernat @fmaussion it's fine to merge your own PR once it's been reviewed & approved by another core developer.

**Comment by shoyer** [comment]
This looks good so let's get it in!

**Comment by rabernat** [comment]
@fmaussion - so sorry I accidentally deleted your tests. I was just rushing to resolve the conflicts so as not to hold up the other PRs. I hope this does not cause too much inconvenience.

**Comment by fmaussion** [comment]
@rabernat no problem, will go back the rasterio tests soon

**Pull request merged**